### PR TITLE
[Config] Introduce concept of deprecating nodes

### PR DESCRIPTION
**Pull request by ro0NL**
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | not updated
| Fixed tickets | #... 
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

This introduces an out-the-box solution for deprecating nodes in a configuration tree, avoid doing it user-land and repetitively.

It works like

```php
// Configuration.php

$someArrayNode
    // trigger notice if given, continue as usual
    ->deprecate(['foo'])

    // trigger single notice if at least one is given
    // specifying a handler means these keys are unset afterwards
    ->deprecate(['bar', 'baz'], function (array $v) { 
        $v['new_bar_baz'] = ($v['bar'] ?? '') . ($v['baz'] ?? '');

        return $v;
    })

    // custom message
    ->deprecate(['a'], null, 'Use "foo" instead')

    // continue as usual
    ->childeren()
        ->scalarNode('foo')->end()
         ->scalarNode('new_bar_baz')->end()
    ->end();
```

Child nodes must be referenced once in mulitple deprecations. meaning you cant do

```php
$array->deprecate(['a']);
$array->deprecate(['a', 'b']);
```

## PR Goal

Utilize (fluent interface) the implementation of `trigger_error($message, E_USER_DEPRECATED)`. So that users only have to care about business value; bridging the old scenario to the new one. Aiming to provide consistent deprecation messages by default (again, to not bother users with it). And doing it rather fast.

The bridging scenario is either a simple case or a complex one, it doesnt differ when doing it with `beforeNormalization` and is not what this PR is about nor aiming to solve. Although simple scenarios can be utilized.

### CODE REVIEW

File: src/Symfony/Component/Config/Definition/ArrayNode.php [hunk 1]
```diff
@@ -52,17 +53,41 @@ public function setNormalizeKeys($normalizeKeys)
...
-                $normalized[$normalizedKey] = $v;
-            } else {
-                $normalized[$k] = $v;
+            if (null === $message) {
+                $message = sprintf(1 === count($given) ? 'The node "%s" is deprecated at path "%s".' : 'The nodes "%s" are deprecated at path "%s".',  implode('", "', $given), $this->getPath());
+            }
+            @trigger_error($message, E_USER_DEPRECATED);
+
+            if (null !== $handler) {
+                $normalized = $handler($this->remapXml($normalized));
```
Comment (ro0NL) [inline src/Symfony/Component/Config/Definition/ArrayNode.php]: Oh, and on a side note; it was a really weird experience noticing that xml remapping is applied after any normalization.

I totally expected this to be done _before_ normalization. Like normalizing keys...

**Comment by ro0NL** [comment]
Re-considered some things, imo.;

- a deprecated node can never be referenced in child nodes
- a deprecation should always have a handler
- a deprecation always unsets the old keys

This makes the behavior more consistent imo. If you only want a notice and preserve nodes you can do that yourself easily.

**Comment by ro0NL** [comment]
Im even thinking expression builder now..

What about

```php
$array
  ->deprecate()
    ifTrue($callable)
    //ifGiven(['foo', 'bar'])
  ->end()
```

**Comment by GuilhemN** [comment]
I wonder if this is worth it, isn't it simple enough to do:
```php
$node
    ->beforeNormalization()
        ->always(function ($v) {
            if (isset($v['foo']) {
                @trigger_error('Foo is deprecated', E_USER_DEPRECATED);
            }
        })
    ->end();
```
?

**Comment by ro0NL** [comment]
Yes and no.

It's simple, but cumbersome. The strategy is always the same here (bridging old values to new ones, removing the old ones).

And im not 100% sure yet, but i believe we should use `array_key_exists` to truly check if a user passed a certain node.

So imagine, deprecating multiple fields... we need checks like `array_key_exists('foo', $v) || array_key_exists('bar', $v)` and `unset()` calls in user land, trying to keep deprecation messages consistent.

Anyway.. without #21052 this approach (nor `beforeNormalization()`) is not waterproof anyway.

**Comment by ro0NL** [comment]
@GuilhemN updated PR description with an explicit goal, why i think it's important.

**Comment by nicolas-grekas** [comment]
I think I'm :-1: here because it adds a feature that is quite easily implemented directly in the code, and that is not needed by userland (well, could be for bundle authors, but that's quite rare anyway.)

**Comment by ro0NL** [comment]
Imo. it could help SF a lot, for contributors to do it intuitively themselves, saving time on reviews. And indeed may help other libs.

From a pure component perspective it may benefit many. At least, that were my considerations in the first place.

**Comment by ro0NL** [comment]
```
$node->deprecate(['foo', 'bar'])
   ->since('1.0')
   ->renameTo('new_%s') // or closure
->end()
```

I prefer that by all means over writing the same type of normalization closures all the time. Deprecations should not be error prone, we standardized it for DI services as well.

**Comment by GuilhemN** [comment]
@ro0nl I understand the goal but I agree with @nicolas-grekas, it is too elaborated for its uses.
It is not common to deprecate a config option, I don't think it is worth dedicated a part of the component to it.

Maybe a simple utility method based on the current features would be enough, without being a nightmare to maintain ?
```php
public function deprecate(string[] $keys, string $message = 'Option "%s" is deprecated.')
{
    return $this->beforeNormalization()
        ->always(function ($v) {
            foreach ($keys as $key) {
                if (array_key_exists($v, $key)) {
                     @trigger_error(sprintf($message, $key), E_USER_DEPRECATED);
                }
            }
            
            return $v;
        })
    ->end();
}

**Comment by ro0NL** [comment]
> It is not common to deprecate a config option

Agree

>  I don't think it is worth dedicated a part of the component to it.

Here's where we disagree. I think the _component_ could solve a common _case_ here. My project has few deprecations, many projects have many deprecations.

Or at least i see a opportunity for the component to provide a standard here. Deprecations in configs are a real issue, and this could aim for better config management and best practices.

Now, i dont want to push anything here and i understand the motivation against it. Always happy to close if i cant convince :)

**Comment by nicolas-grekas** [comment]
In the hope to save everyones' time, I'm :-1: here :)

**Comment by fabpot** [comment]
I don't see the need for end user projects to deprecate options, so I'm 👎 as well

**Comment by ro0NL** [comment]
Well.. then it's 4-4.. making me already less convinced :)

Hope the point is clear this is not about adding features, for the sake of it. The case is tedious, and the component goes beyond "end user projects".

:+1:

**Pull request closed without merging**